File: src/main/java/org/simbrain/custom_sims/simulations/sorn/AddSTDPRule.java
package org.simbrain.custom_sims.simulations.sorn;

import org.simbrain.network.core.Synapse;
import org.simbrain.network.core.SynapseUpdateRule;
import org.simbrain.network.synapse_update_rules.STDPRule;
import org.simbrain.network.util.ScalarDataHolder;

public class AddSTDPRule extends STDPRule {

    private boolean srcSpk = false;

    private boolean tarSpk = false;

    @Override
    public void init(Synapse synapse) {
    }

    @Override
    public String getName() {
        return "STDP";
    }

    @Override
    public SynapseUpdateRule deepCopy() {
        STDPRule duplicateSynapse = new STDPRule();
        duplicateSynapse.setTau_minus(this.getTau_minus());
        duplicateSynapse.setTau_plus(this.getTau_plus());
        duplicateSynapse.setW_minus(this.getW_minus());
        duplicateSynapse.setW_plus(this.getW_plus());
        duplicateSynapse.setLearningRate(this.getLearningRate());
        return duplicateSynapse;
    }

    @Override
    public void apply(Synapse synapse, ScalarDataHolder data) {
        boolean ss = synapse.getSource().isSpike();
        boolean st = synapse.getTarget().isSpike();

        double str = synapse.getStrength();

        str += learningRate * ((srcSpk && st ? 1 : 0) - (tarSpk && ss ? 1 : 0));
        synapse.setStrength(synapse.clip(str));
        srcSpk = ss;
        tarSpk = st;

    }
}

File: src/main/kotlin/org/simbrain/custom_sims/simulations/sorn/AddSTDPRule.kt
package org.simbrain.custom_sims.simulations.sorn

import org.simbrain.network.core.Network
import org.simbrain.network.core.Synapse
import org.simbrain.network.learningrules.STDPRule
import org.simbrain.network.learningrules.SynapseUpdateRule
import org.simbrain.network.util.EmptyScalarData

class AddSTDPRule : STDPRule() {
    private var srcSpk = false

    private var tarSpk = false

    override fun init(synapse: Synapse) {
    }

    override val name: String
        get() = "STDP"

    override fun copy(): SynapseUpdateRule<*, *> {
        val duplicateSynapse = STDPRule()
        duplicateSynapse.tauMinus = this.tauMinus
        duplicateSynapse.tauPlus = this.tauPlus
        duplicateSynapse.wMinus = this.wMinus
        duplicateSynapse.wPlus = this.wPlus
        duplicateSynapse.learningRate = this.learningRate
        return duplicateSynapse
    }

    context(Network)
    override fun apply(synapse: Synapse, data: EmptyScalarData) {
        val ss = synapse.source.isSpike
        val st = synapse.target.isSpike
        synapse.strength += learningRate * ((if (srcSpk && st) 1 else 0) - (if (tarSpk && ss) 1 else 0))
        srcSpk = ss
        tarSpk = st
    }
}
